File: lib/features/reservation/domain/entities/vehicle.dart
import 'package:equatable/equatable.dart';

/// Entity représentant un véhicule
class Vehicle extends Equatable {
  final String id;
  final String userId;
  final String make; // Marque (ex: Honda)
  final String model; // Modèle (ex: Civic)
  final int? year; // Année
  final String color; // Couleur
  final String? licensePlate;
  final String? photoUrl; // Photo du véhicule
  final VehicleType type;
  final bool isDefault; // Véhicule par défaut de l'utilisateur
  final DateTime createdAt;
  final DateTime updatedAt;

  const Vehicle({
    required this.id,
    required this.userId,
    required this.make,
    required this.model,
    this.year,
    required this.color,
    this.licensePlate,
    this.photoUrl,
    this.type = VehicleType.car,
    this.isDefault = false,
    required this.createdAt,
    required this.updatedAt,
  });

  /// Affichage formaté du véhicule
  String get displayName {
    if (year != null) {
      return '$make $model $year';
    }
    return '$make $model';
  }

  /// Affichage court (pour les listes)
  String get shortName {
    return '$make $model';
  }

  /// Affichage avec couleur
  String get displayWithColor {
    return '$color $make $model';
  }

  Vehicle copyWith({
    String? id,
    String? userId,
    String? make,
    String? model,
    int? year,
    String? color,
    String? licensePlate,
    String? photoUrl,
    VehicleType? type,
    bool? isDefault,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return Vehicle(
      id: id ?? this.id,
      userId: userId ?? this.userId,
      make: make ?? this.make,
      model: model ?? this.model,
      year: year ?? this.year,
      color: color ?? this.color,
      licensePlate: licensePlate ?? this.licensePlate,
      photoUrl: photoUrl ?? this.photoUrl,
      type: type ?? this.type,
      isDefault: isDefault ?? this.isDefault,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }

  @override
  List<Object?> get props => [
        id,
        userId,
        make,
        model,
        year,
        color,
        licensePlate,
        photoUrl,
        type,
        isDefault,
        createdAt,
        updatedAt,
      ];
}

/// Types de véhicules
enum VehicleType {
  car, // Voiture
  suv, // VUS
  truck, // Camion
  van, // Fourgonnette
  motorcycle, // Moto (été)
}

extension VehicleTypeExtension on VehicleType {
  String get displayName {
    switch (this) {
      case VehicleType.car:
        return 'Voiture';
      case VehicleType.suv:
        return 'VUS';
      case VehicleType.truck:
        return 'Camion';
      case VehicleType.van:
        return 'Fourgonnette';
      case VehicleType.motorcycle:
        return 'Moto';
    }
  }

  String get icon {
    switch (this) {
      case VehicleType.car:
        return '🚗';
      case VehicleType.suv:
        return '🚙';
      case VehicleType.truck:
        return '🚚';
      case VehicleType.van:
        return '🚐';
      case VehicleType.motorcycle:
        return '🏍️';
    }
  }

  /// Multiplicateur de prix basé sur le type
  double get priceFactor {
    switch (this) {
      case VehicleType.car:
        return 1.0;
      case VehicleType.suv:
        return 1.2;
      case VehicleType.truck:
        return 1.3;
      case VehicleType.van:
        return 1.25;
      case VehicleType.motorcycle:
        return 0.7;
    }
  }
}

File: backend/models/Vehicle.js
/**
 * Modèle Mongoose pour les véhicules des clients.
 * Stocke les informations du véhicule (marque, modèle, plaque) et le véhicule par défaut.
 */

const mongoose = require('mongoose');

// --- Schéma principal ---

const vehicleSchema = new mongoose.Schema({
    userId: {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'User',
        required: [true, 'L\'utilisateur est requis'],
    },
    make: {
        type: String,
        required: [true, 'La marque est requise'],
        trim: true,
    },
    model: {
        type: String,
        required: [true, 'Le modèle est requis'],
        trim: true,
    },
    year: {
        type: Number,
        required: [true, 'L\'année est requise'],
        min: [1900, 'Année invalide'],
        max: [new Date().getFullYear() + 1, 'Année invalide'],
    },
    color: {
        type: String,
        required: [true, 'La couleur est requise'],
        trim: true,
    },
    licensePlate: {
        type: String,
        required: [true, 'La plaque d\'immatriculation est requise'],
        trim: true,
        uppercase: true,
    },
    type: {
        type: String,
        enum: ['sedan', 'suv', 'truck', 'van', 'coupe', 'hatchback', 'car'],
        required: [true, 'Le type de véhicule est requis'],
        default: 'sedan',
    },
    photoUrl: {
        type: String,
        default: null,
    },
    isDefault: {
        type: Boolean,
        default: false,
    },
    isActive: {
        type: Boolean,
        default: true,
    },
}, {
    timestamps: true,
});

// --- Index ---

vehicleSchema.index({ userId: 1, isActive: 1 });
vehicleSchema.index({ licensePlate: 1 });

// --- Propriétés virtuelles ---

vehicleSchema.virtual('displayName').get(function() {
    return `${this.make} ${this.model} (${this.year})`;
});

// Méthode pour obtenir l'émoji selon le type
vehicleSchema.virtual('icon').get(function() {
    const icons = {
        sedan: '🚗',
        suv: '🚙',
        truck: '🛻',
        van: '🚐',
        coupe: '🏎️',
        hatchback: '🚗',
        car: '🚗',

    };
    return icons[this.type] || '🚗';
});

// --- Middleware pre-save ---

// S'assure qu'un seul véhicule par défaut existe par utilisateur
vehicleSchema.pre('save', async function(next) {
    if (this.isDefault && this.isModified('isDefault')) {
        await this.constructor.updateMany(
            { userId: this.userId, _id: { $ne: this._id } },
            { isDefault: false }
        );
    }
    next();
});

// Options pour inclure les virtuals dans JSON
vehicleSchema.set('toJSON', { virtuals: true });
vehicleSchema.set('toObject', { virtuals: true });

module.exports = mongoose.model('Vehicle', vehicleSchema);